src/librbd/ImageWatcher.cc

```diff
@@ -816,10 +819,23 @@ Context *ImageWatcher<I>::prepare_quiesce_request(
       std::unique_lock async_request_locker{m_async_request_lock};
       mark_async_request_complete(request, r);
       auto ctx = remove_async_request(request, m_async_request_lock);
-      ceph_assert(ctx != nullptr);
-      ctx = new C_ResponseMessage(static_cast<C_NotifyAck *>(ctx));
+      if (ctx != nullptr) {
+        ctx = new C_ResponseMessage(static_cast<C_NotifyAck *>(ctx));
+      }
       async_request_locker.unlock();
-      ctx->complete(r);
+      if (ctx != nullptr) {
+        ctx->complete(r);
+      } else { // canceled
+        bool canceled = m_task_finisher->cancel(Task(TASK_CODE_QUIESCE, request));
```

Comment:
Is this trying to cover a potential race where the image is closed while a quiesce request is still being processed? Perhaps if `TaskFinisher::cancel_event` actually fired the callback contexts w/ `-ECANCELED`, this code could be avoided since `unquiesce_ctx` would be re-used.

Reply:
Do you mean adding a new `TaskFinisher::cancel_event` method with this behavior and use it here? Because I think we can't modify the existing `TaskFinisher::cancel` method as in other cases we want the context is just deleted.

Reply:
No, I'm saying tweak `TaskFinisher::cancel` to invoke `it->second.first->complete(-ECANCELED)` and update the `ImageWatcher` task callback functions to ignore any callbacks with `r == -ECANCELED` (except for the quiesce one).

Reply:
It will change change a behavior a little for the case when `unquiesce` notification is recieved (see `prepare_unquiesce_request`). Right now it cancels the scheduled `unquiesce` and runs its own, returning the ack message only after it is complete.

After I change `TaskFinisher::cancel` I have to send the ack not waiting for `unquiesce` to complete, which probably is also ok. Just want to make sure you are ok with this too...

Reply:
Ah. never mind. Actually we already have this behavior. Ok, I will try this way.